### DISCUSSION TITLE
maxNumberOfThreads property usage

### DISCUSSION THREAD

**Discussion by c-ankur** [americanexpress/unify-flowret#21] [General] [answered]
Just would like to confirm that the property maxNumberOfThreads is used only for parallel execution scenarios?

Or does it also determine how many max number of flowret cases can run at a given moment?
Is there a limit to how many parallel cases can run in flowret? Is there a way to define a thread pool max threads for this?

**Comment by deepakarora3** [comment, marked as answer]
The maxNumberOfThreads is only used for parallel execution. The number specified here will be used across all parallel executions. For example, if there are 15 concurrent executions of parallel processing, then those 15 will share the number of threads specified in this property. In this sense, it is a pool of threads that are shared for all parallel executions. No, there is no limit to the number of Flowret cases that can be run at any given moment. It will depend on the JVM CPU and memory and in case of parallel processing on the value of this parameter. Hope this clarifies.

**Reply by c-ankur** [reply]
Thanks.
To clarify with an example: If each case has 2 parallel branches/tasks, and there are 15 such flowret cases started, and the maxNumberOfThreads property is set to 2:-
a) will flowret allow only, either 1 case with 2 tasks, or 2 cases with 1 task, to run in parallel?
b) or will flowret allow all 15 cases to run with 2 parallel tasks each, at the same time?

From your answer above, I am assuming 
- only a) is TRUE.
- In order to run all 15 cases in parallel, the maxNumberOfThreads should be set to a higher number, let's say 30 (15 x 2).
- This means that the maxNumberOfThreads is applied on the branch level (rather than on the case level)

Please correct me if wrong.

**Reply by deepakarora3** [reply]
In the example you have outlined, at most only two threads will run in parallel and these two threads can be from any case. Basically, which ever case gets the thread will start executing and other cases will wait till the time a thread is made available. I think that I have set the thread scheduling policy to be fair which means that the threads will be allocated to the cases in the order in which they were requested, and this is to avoid a situation where one case waits forever. Hope this clarifies.
